gqlalchemy/vendors/database.py

```diff
@@ -0,0 +1,316 @@
+        """Creates a constraint (label or label-property type) in the database."""
+        query = f"CREATE CONSTRAINT ON {index.to_cypher()};"
+        self.execute(query)
+
+    def drop_constraint(self, index: Constraint) -> None:
+        """Drops a constraint (label or label-property type) in the database."""
+        query = f"DROP CONSTRAINT ON {index.to_cypher()};"
+        self.execute(query)
+
+    @abstractmethod
+    def get_constraints(
+        self,
+    ) -> List[Constraint]:
+        """Returns a list of all database constraints (label and label-property types)."""
+        pass
+
+    @abstractmethod
+    def get_exists_constraints(
+        self,
+    ) -> List[Constraint]:
+        pass
+
+    @abstractmethod
+    def get_unique_constraints(
+        self,
+    ) -> List[Constraint]:
+        pass
+
+    def ensure_constraints(
+        self,
+        constraints: List[Constraint],
+    ) -> None:
+        """Ensures that database constraints match input constraints."""
+        old_constraints = set(self.get_constraints())
+        new_constraints = set(constraints)
+        for obsolete_constraints in old_constraints.difference(new_constraints):
+            self.drop_constraint(obsolete_constraints)
+        for missing_constraint in new_constraints.difference(old_constraints):
+            self.create_constraint(missing_constraint)
+
+    def drop_database(self):
+        """Drops database by removing all nodes and edges."""
+        self.execute("MATCH (n) DETACH DELETE n;")
+
+    def _get_cached_connection(self) -> Connection:
+        """Returns cached connection if it exists, creates it otherwise."""
+        if self._cached_connection is None or not self._cached_connection.is_active():
+            self._cached_connection = self.new_connection()
+
+        return self._cached_connection
+
+    @abstractmethod
+    def new_connection(self) -> Connection:
+        """Creates new database connection."""
+        pass
+
+    def _get_nodes_with_unique_fields(self, node: Node) -> Optional[Node]:
+        """Get's all nodes from the database that have any of the unique fields
+        set to the values in the `node` object.
+        """
+        return self.execute_and_fetch(
+            f"MATCH (node: {node._label})"
+            + f" WHERE {node._get_cypher_unique_fields_or_block('node')}"
+            + " RETURN node;"
+        )
+
+    def get_variable_assume_one(self, query_result: Iterator[Dict[str, Any]], variable_name: str) -> Any:
+        """Returns a single result from the query_result (usually gotten from
+        the execute_and_fetch function).
+        If there is more than one result, raises a GQLAlchemyError.
+        """
+        result = next(query_result, None)
+        next_result = next(query_result, None)
+        if result is None:
+            raise GQLAlchemyError("No result found. Result list is empty.")
+        elif next_result is not None:
```

Comment:
consider switching these elifs to simple ifs. Every branch raises, so it makes sense